eth/vm/logic/storage.py

```diff
@@ -49,6 +49,71 @@ def sstore(computation):
     )
 
 
+def sstore_eip1283(computation):
+    slot, value = computation.stack_pop(num_items=2, type_hint=constants.UINT256)
+
+    current_value = computation.state.account_db.get_storage(
+        address=computation.msg.storage_address,
+        slot=slot,
+    )
+
+    original_value = computation.state.account_db.get_storage(
+        address=computation.msg.storage_address,
+        slot=slot,
+        original=True
+    )
+
+    is_currently_empty = not bool(current_value)
+    is_going_to_be_empty = not bool(value)
+
+    gas_refund = 0
+
+    if current_value == value:
```

Comment:
This whole block looks like an excellent candidate for a stand-alone utility function which can be tested in isolation.

Also, if we convert all of the inputs to booleans:

```python
def get_gas_cost_and_refund(original_is_zero, current_is_zero, value_is_zero, original_equals_current, original_equals_value):
    ...
```

We end up with a function that has `2**5 -> 32` possible input combinations for which we should be able to explicitely specify what the expected costs are.

Reply:
Unfortunately though, following the spec causes negative refunds for some configurations for us. I believe the reason for this is:

>If an implementation uses “execution-frame level” refund counter (a new refund counter is created at each call frame, and then merged back to parent when the call frame finishes), then the refund counter needs to be changed to signed – at internal calls, a child refund can go below zero.

This seems to apply to us. So, if you refund counter would be a single instance *per transaction*, then we would not get negative refunds, but since our refund counter is per Computation/Message, we get negatives.

That doesn't contradict what you wrote, we could just assert getting negative results for some of the combinations. I still have to taker a closer look to see what would be the best way forward.

Reply:
I think that implies that our `GasMeter` may need to be changed since it doesn't allow for negative refunds.  And likely we just need to create a new `ConstantinopleGasMeter` since the logic isn't needed for pre-constantinople forks.

Reply:
I'm trying out something more compositional right now but I'm not sure if I like it because I think it doesn't play so well with the rest of our architecture (e.g. we do not use a DI framework which takes the boilerplate out of composition). So I may fall back to what you suggest but I want to give it a chance at least.